sqlmesh/core/engine_adapter/athena.py

```diff
@@ -373,6 +409,101 @@ def _find_matching_columns(
                 matches.append((key, match_dtype))
         return matches
 
+    def _insert_overwrite_by_time_partition(
+        self,
+        table_name: TableName,
+        source_queries: t.List[SourceQuery],
+        columns_to_types: t.Dict[str, exp.DataType],
+        where: exp.Condition,
+        **kwargs: t.Any,
+    ) -> None:
+        table_type = self._query_table_type(table_name)
+
+        if table_type == "iceberg":
+            # Iceberg tables work as expected, we can use the default behaviour
+            return super()._insert_overwrite_by_time_partition(
+                table_name, source_queries, columns_to_types, where, **kwargs
+            )
+
+        # For Hive tables, we need to drop all the partitions covered by the query and delete the data from S3
+        self._clear_partition_data(table_name, where)
+
+        # Now the data is physically gone, we can continue with inserting a new partition
+        return super()._insert_overwrite_by_time_partition(
+            table_name,
+            source_queries,
+            columns_to_types,
+            where,
+            insert_overwrite_strategy_override=InsertOverwriteStrategy.INTO_IS_OVERWRITE,  # since we already cleared the data
+            **kwargs,
+        )
+
+    def _clear_partition_data(self, table: TableName, where: t.Optional[exp.Condition]) -> None:
```

Comment:
table -> table_name

Reply:
> or what you should do is clean up so all internal methods take in exp.Table instead

Unfortunately a bunch are defined in the superclass which means they cant be easily cleaned up. I'll clean up the ones that only exist in this adapter class though